### PR TITLE
Ccip-3398 populate state initial PR

### PR DESCRIPTION
**Pull request by AnieeG**
Adds snapshot creation for a set of contracts. The state snapshot can be updated later based on what more fields are required for each one. This is just to set the standard approach to be followed for state extraction.

- TokenAdminRegistry         
- FeeQuoter                    
- NonceManager                 
- Router                       
- RMN                          
- OnRamp  

Rest of the contracts will be covered on subsequent PRs.

### CODE REVIEW

File: integration-tests/deployment/ccip/state.go [hunk 1]
```diff
@@ -219,7 +213,7 @@ func LoadChainState(chain deployment.Chain, addresses map[string]deployment.Type
 			if err != nil {
 				return state, err
 			}
-			state.RMNRemote = rmnRemote
+			state.RMNRemote = rmn1_6.New(rmnRemote)
```
Comment (ogtownsend) [inline integration-tests/deployment/ccip/state.go]: how come `rmn1_6` if this case statement is pointing to `Version1_0_0`?

```
case deployment.NewTypeAndVersion(RMNRemote, deployment.Version1_0_0...
```

Comment (AnieeG) [reply integration-tests/deployment/ccip/state.go]: good catch, I will update it.

Comment (AnieeG) [reply integration-tests/deployment/ccip/state.go]: @ogtownsend can you point me to the code ref ? I see that TypeAndVersion for RMN is [RMNRemote 1.6.0-dev](https://github.com/smartcontractkit/chainlink/blob/develop/contracts/src/v0.8/ccip/rmn/RMNRemote.sol#L17)

Comment (AnieeG) [reply integration-tests/deployment/ccip/state.go]: I see what you mean, @connorwstein I think the case statement should be changed to `Version1_6_0_dev`, am I missing something?

Comment (connorwstein) [reply integration-tests/deployment/ccip/state.go]: yeah should be changed

File: integration-tests/deployment/ccip/view/chain.go [hunk 2]
```diff
@@ -0,0 +1,31 @@
+package view
+
+import (
+	"github.com/smartcontractkit/chainlink/integration-tests/deployment/ccip/view/types/v1_2"
+	"github.com/smartcontractkit/chainlink/integration-tests/deployment/ccip/view/types/v1_5"
+	"github.com/smartcontractkit/chainlink/integration-tests/deployment/ccip/view/types/v1_6"
+)
+
+type Chain struct {
+	DestinationChainSelectors []uint64 `json:"destinationChainSelectors,omitempty"`
```
Comment (connorwstein) [inline integration-tests/deployment/ccip/view/chain.go]: I worry about caching state from specific contracts at this top level - hard to follow where it comes from. Can we keep the state from each contract in the contract specific view so that its explicit?

File: integration-tests/deployment/ccip/view/chain.go [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+	// TODO - populate supportedTokensByDestination
+	SupportedTokensByDestination map[uint64][]string                `json:"supportedTokensByDestination,omitempty"`
+	TokenAdminRegistry           map[string]v1_5.TokenAdminRegistry `json:"tokenAdminRegistry,omitempty"`
```
Comment (connorwstein) [inline integration-tests/deployment/ccip/view/chain.go]: interesting - I wasn't imagining we'd have multiple deployments of the same version of a contract on the same chain at any given time, but I guess its still possible so good to keep that flexibility.

File: integration-tests/deployment/ccip/view/chain.go [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+	FeeQuoter                    map[string]v1_6.FeeQuoter          `json:"feeQuoter,omitempty"`
```
Comment (connorwstein) [inline integration-tests/deployment/ccip/view/chain.go]: I think the json marshalling would still work if we kept keys as common.Address right because they have a String() method? Stronger typing there would prevent unexpected keys like "blah"

File: integration-tests/deployment/ccip/view/types/contract_state.go [hunk 5]
```diff
@@ -0,0 +1,33 @@
+package types
```
Comment (connorwstein) [inline integration-tests/deployment/ccip/view/types/contract_state.go:1]: idk if the types directory is helping much - think we could just have 
```
view/chain.go
view/v1_2/..
view/v1_5/...
```

File: integration-tests/deployment/ccip/view/types/contract_state.go [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+	"decommissioning": Decommissioning,
+	"dead":            Dead,
+}
+
+type Contract struct {
+	TypeAndVersion string `json:"typeAndVersion,omitempty"`
+	Address        string `json:"address,omitempty"`
+}
+
+type ContractState interface {
```
Comment (connorwstein) [inline integration-tests/deployment/ccip/view/types/contract_state.go]: unused interface?

File: integration-tests/deployment/ccip/view/types/contract_state.go [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+var ContractStatusLookup = map[string]ContractStatus{
+	"active":          Active,
+	"inactive":        Inactive,
+	"decommissioning": Decommissioning,
+	"dead":            Dead,
+}
+
+type Contract struct {
+	TypeAndVersion string `json:"typeAndVersion,omitempty"`
+	Address        string `json:"address,omitempty"`
```
Comment (connorwstein) [inline integration-tests/deployment/ccip/view/types/contract_state.go]: better to leave this as an common.Address?

File: integration-tests/deployment/ccip/view/types/v1_2/router.go [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+
+type Router struct {
+	types.Contract
+	WrappedNative string            `json:"wrappedNative,omitempty"`
+	ARMProxy      string            `json:"armProxy,omitempty"`
+	OnRamps       map[uint64]string `json:"onRamps,omitempty"`  // Map of DestinationChainSelectors to OnRamp Addresses
+	OffRamps      map[uint64]string `json:"offRamps,omitempty"` // Map of SourceChainSelectors to a list of OffRamp Addresses
+}
+
+func (r Router) Address() common.Address {
```
Comment (connorwstein) [inline integration-tests/deployment/ccip/view/types/v1_2/router.go]: I think we can avoid this boilerplate if we just leave type.Contract.Address as a common.Address

File: integration-tests/deployment/ccip/view/types/v1_2/router.go [hunk 9]
```diff
@@ -0,0 +1,69 @@
+package v1_2
+
+import (
+	"github.com/ethereum/go-ethereum/common"
+
+	"github.com/smartcontractkit/chainlink/integration-tests/deployment/ccip/view/types"
+	"github.com/smartcontractkit/chainlink/v2/core/gethwrappers/ccip/generated/router"
+)
+
+type Router struct {
```
Comment (connorwstein) [inline integration-tests/deployment/ccip/view/types/v1_2/router.go]: might be better to call this RouterView for clarity? Then all the snapshot methods would look something like `XSnapshot(gethwrapper for x) XView` which makes sense to me.

File: integration-tests/deployment/ccip/view/types/v1_2/router.go [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+		return Router{}, err
+	}
+	onRamps := make(map[uint64]string)
+	offRamps := make(map[uint64]string)
+	offRampList, err := r.GetOffRamps(nil)
+	if err != nil {
+		return Router{}, err
+	}
+	for _, offRamp := range offRampList {
+		offRamps[offRamp.SourceChainSelector] = offRamp.OffRamp.Hex()
```
Comment (connorwstein) [inline integration-tests/deployment/ccip/view/types/v1_2/router.go]: probably String() more readable/natural vs Hex

File: integration-tests/deployment/ccip/view/types/v1_2/router.go [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+
+	"github.com/smartcontractkit/chainlink/integration-tests/deployment/ccip/view/types"
+	"github.com/smartcontractkit/chainlink/v2/core/gethwrappers/ccip/generated/router"
+)
+
+type Router struct {
+	types.Contract
+	WrappedNative string            `json:"wrappedNative,omitempty"`
+	ARMProxy      string            `json:"armProxy,omitempty"`
+	OnRamps       map[uint64]string `json:"onRamps,omitempty"`  // Map of DestinationChainSelectors to OnRamp Addresses
```
Comment (connorwstein) [inline integration-tests/deployment/ccip/view/types/v1_2/router.go]: Since the views are supposed to be human readable, I think we want to resolve the chain selectors into chain names here

File: integration-tests/deployment/ccip/view/types/v1_5/tokenadminregistry.go [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+	types.Contract
+	Tokens []common.Address `json:"tokens"`
+}
+
+func (ta TokenAdminRegistry) Address() common.Address {
+	return common.HexToAddress(ta.Contract.Address)
+}
+
+func TokenAdminRegistrySnapshot(taContract *token_admin_registry.TokenAdminRegistry) (TokenAdminRegistry, error) {
+	tokens, err := taContract.GetAllConfiguredTokens(nil, 0, 10)
```
Comment (connorwstein) [inline integration-tests/deployment/ccip/view/types/v1_5/tokenadminregistry.go]: will be an interesting problem here, there could be an unbounded number of tokens since its permissionless. We can start by just listing them all but lets add a TODO here for smarter pruning/DoS protection later.

File: integration-tests/deployment/ccip/view/types/v1_2/router.go [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+	}
+	for selector := range offRamps {
+		onRamp, err := r.GetOnRamp(nil, selector)
+		if err != nil {
+			return Router{}, err
+		}
+		onRamps[selector] = onRamp.Hex()
+	}
+	return Router{
+		Contract: types.Contract{
```
Comment (connorwstein) [inline integration-tests/deployment/ccip/view/types/v1_2/router.go]: Owner is a pretty widespread field we can include that in the Contract type too I think

File: integration-tests/deployment/ccip/view/types/v1_5/tokenadminregistry.go [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+	if err != nil {
+		return TokenAdminRegistry{}, err
+	}
+	tv, err := taContract.TypeAndVersion(nil)
+	if err != nil {
+		return TokenAdminRegistry{}, err
+	}
+	return TokenAdminRegistry{
```
Comment (connorwstein) [inline integration-tests/deployment/ccip/view/types/v1_5/tokenadminregistry.go]: we definitely want to display the registry modules too. Sadly we have to filter events for that

File: integration-tests/deployment/ccip/view/types/v1_6/feequoter.go [hunk 15]
```diff
@@ -0,0 +1,129 @@
...
+	EnforceOutOfOrder                 bool   `json:"enforceOutOfOrder,omitempty"`
+	ChainFamilySelector               string `json:"chainFamilySelector,omitempty"`
+}
+
+type FeeQuoterTokenPriceFeedConfig struct {
+	DataFeedAddress string `json:"dataFeedAddress,omitempty"`
+	TokenDecimals   uint8  `json:"tokenDecimals,omitempty"`
+}
+
+func FeeQuoterSnapshot(fqContract *fee_quoter.FeeQuoter, tokensByDestination map[uint64][]string) (FeeQuoter, error) {
```
Comment (connorwstein) [inline integration-tests/deployment/ccip/view/types/v1_6/feequoter.go]: I'd prefer just only gethwrapper arguments, makes it explicit which contracts the view has dependencies on

File: integration-tests/deployment/ccip/view/types/contract_state.go (hunk 5 above)
Comment (AnieeG) [reply integration-tests/deployment/ccip/view/types/contract_state.go:1]: The type directory is to avoid import cycle, We have the Contract struct getting imported into v1_2 ,v1_5 etc from parent view and then Chain struct has elements from v1_2, v1_5 in the parent view package

File: integration-tests/deployment/ccip/view/types/v1_2/router.go (hunk 11 above)
Comment (AnieeG) [reply integration-tests/deployment/ccip/view/types/v1_2/router.go]: Chain name derived from chain-selectors repo?

File: integration-tests/deployment/ccip/view/chain.go (hunk 2 above)
Comment (AnieeG) [reply integration-tests/deployment/ccip/view/chain.go]: Needed DestinationChainSelectors in `OnRamp` and `feeQuoter` as of now to populate corresponding config. Do you suggest we pass router struct in the corresponding snapshot function to retrieve the chain selectors by calling getOffRamps?

File: integration-tests/deployment/ccip/view/types/v1_2/router.go (hunk 11 above)
Comment (connorwstein) [reply integration-tests/deployment/ccip/view/types/v1_2/router.go]: on second thought, let's think more about how to do this in a uniform way for both selectors and addresses. Can leave it like this for now

File: integration-tests/deployment/ccip/view/types/v1_2/router.go (hunk 10 above)
Comment (AnieeG) [reply integration-tests/deployment/ccip/view/types/v1_2/router.go]: I will change it to common.Address to keep it uniform for all addresses

File: integration-tests/deployment/ccip/view/types/contract_state.go (hunk 5 above)
Comment (connorwstein) [reply integration-tests/deployment/ccip/view/types/contract_state.go:1]: I see - I think it could be on the same hierarchical level though just to avoid the nesting

File: integration-tests/deployment/ccip/view/types/contract_state.go [hunk 16]
```diff
@@ -0,0 +1,33 @@
...
+)
+
+type ContractMetaData struct {
+	TypeAndVersion string         `json:"typeAndVersion,omitempty"`
+	Address        common.Address `json:"address,omitempty"`
+	Owner          common.Address `json:"owner,omitempty"`
+}
+
+func NewContractMetaData(tv Meta, addr common.Address) (ContractMetaData, error) {
+	tvStr, err := tv.TypeAndVersion(nil)
```
Comment (connorwstein) [inline integration-tests/deployment/ccip/view/types/contract_state.go:15]: these might not always exist (e.g. WETH) but we can come back to that later, can just be empty in that case

File: integration-tests/deployment/ccip/state.go [hunk 17]
```diff
@@ -49,6 +54,63 @@ type CCIPChainState struct {
 	TestRouter *router.Router
 }
 
+func (c CCIPChainState) GenerateView() (view.ChainView, error) {
+	chainView := view.NewChain()
+	r := c.Router
```
Comment (connorwstein) [inline integration-tests/deployment/ccip/state.go]: ultranit - we can avoid these variables and just use c.Router then if c.Router != nil {c.Router.Address()}

**Review by connorwstein — DISMISSED** [review]

**Review by connorwstein — DISMISSED** [review]

**Review by connorwstein — APPROVED** [review]

**Review by cmalec — APPROVED** [review]
